### DISCUSSION TITLE
Long-Term Stash Project Planning: Strategy and Growth

### DISCUSSION THREAD

**Discussion by kermieisinthehouse** [stashapp/stash#1800] [General]
# _**To organize the world's porn, and make that data universally accessible and useful**_

We want the Stash project to be a successful, encompassing, and correct source of information about adult media. I am proposing a new mission statement, modeled after [a certain search company's mission statement](https://about.google/): "_To organize the world's porn, and make that data universally accessible and useful._"  This is not an easy task. By its nature, adult media is largely ephemeral, social taboos keep open discussion to a minimum, and producers/distributors are often short-lived, with little incentive to preserve their works.
To this end, I am also proposing three different initiatives that would help us achieve this task, all of which encourage each other to succeed:  **Growing our Userbase**, **Growing our Dataset**, and **Making it Accessible and Useful**. These are large tasks and may take several years to fully realize, but many have low hanging fruit that can be immediately addressed.

![Flywheel](https://user-images.githubusercontent.com/84814418/135774963-2cce7332-d940-4014-8ae5-9fee0d01dfc9.png)

## Growing Our Userbase
Our value proposition is simple: We can provide out-of-the-box easy organization for your adult media collection, and you can optionally provide some of your organization and research back to us.

### Stash new user funnel: Finding and installing the Stash client

- We need to develop a community outreach strategy: Where can we make ourselves known? How can we reach the people who are passionate about organization, archiving, and preservation of adult media?
- Redo Readme.md, feature big calls to action that link to install media
- Simplified installation instructions, de-emphasize or move to separate document developer info
- Graphics and shininess - Screenshots in the readme at the minimum
- Tagline desperately needed. "Create your own tube site", "Organize your adult media collection in minutes", "Self-host your porn collection"? Needs some workshopping here
- Demo page a possibility? A public stash instance with public domain videos that resets the database every 30 minutes?

### Stash new user funnel: First-use, and reduce time to usefulness
Out-of-the-box usefulness improves uptake and user stickiness.
- We should attempt to open the default browser automatically after server start
- Empty / 0 entry scene / image list should have stashes list and scan button
- Empty performers / studios / tags list should have option to auto-populate with starter from StashDB.
- StashDB should be pre-configured with a read-only user out-of-the-box. A user should be able to start up stash and start leveraging our data immediately, without joining our discord and messing with API keys. ("What's an API key? Do I have to pay for that? Plex just told me what my movies were automatically!")

#### Starter Database: A symptom of missing features.
StashDB should have a living set of "Starter" or "Common" data, perhaps implemented by a flag on each Performer, Tag, or Studio that denotes it as a good starter. (Criteria can be settled on, but anything that's extremely common or widely distributed is a good candidate) On first run, Stash should offer to automatically populate these directly from StashDB, and kick-off a scan and auto-tag job immediately after. This would get new users immediately started on a good foundation toward organization and a clear example of Stash's capabilities.

### Make Stash a delight to use
Stash has a few papercuts. After providing support to many, many users, I've distilled the most common sources of confusion:

- "Save" buttons in settings should be obvious if settings have changed!
- The saved filters feature needs improved discoverability. If the current filter set matches a saved filter, show the name of that view, if not, show a "Save as smart filter" button. If the current filter set is not the default, show a "Set as default" button in the top bar.
- Update notifications should be highly visible: An automated update mechanism should be investigated (click to download stash server, replace binary, exec. Docker support to be discussed): security and bug fixes should be adopted quickly, will allow faster and more incremental releases later, allow greater agility for StashDB changes. Avoid fragmenting the install base.

## Growing Our Dataset
Anything we do must respect user privacy, we must act with full knowledge that our domain is adult media that carries a reputational risk to our users. We DON'T want to know WHO are users are, or anything about them; We only want their contributions, annotations, and scene metadata. All data-sharing should be extremely obvious, all submissions to StashDB should be laid out in front of the user before sending.

- Attempt to establish a data-sharing agreement with ThePornDB or a similar project, to bolster crowd-sourced data with scraped data. Their scene data is missing phash, our phashes are missing scene data. Win/win.

- Stash-box should be de-emphasized in docs. Vast majority have no need to know what stash-box is, and people setting up StashDB will be confused. The few people who do understand what it's for further fracture the potential StashDB dataset, and should use StashDB in almost all cases. StashDB should be the default URL, and changing the stash-box API url should be an advanced feature.

- StashDB should allow you to match Tags, Performers, and Studios to their corresponding StashID. Submitting information to StashDB should encompass scene oshashes, phashes, tags, filenames, performers, and studios. 

- Importantly, Markers and their primary tags should be uploaded to StashDB. OsHash/Scene StashID <-> Tag + Start Timestamp + End Timestamp.

- Local annotation should be first-class. We can increase the amount of crowdsourced contributions by making it easier and more enjoyable to manually add Tags, Performers, Studios. A user should be able to type auto-completed tags while watching a scene preview or viewing the scene thumbnail sprite, then save them and move on to the next scene WITHOUT taking their hands off the keyboard.

- A separate view that enumerates pending changes to upload to StashDB, and past contributions under your account will be necessary. Perhaps also a good UI location for any conflict resolution that's needed. 

## Making Our Data Accessible and Useful
Our biggest asset and differentiator is that our database has file hashes and and perceptual hashes for scenes. Our client software can allow us to crowdsource from huge collections of media, and distribute it to a wide userbase.

- We currently encourage users to delete duplicate phashes! We should store and forward phash->oshash relationships so that StashDB can know (and match) all the common oshashes for a particular scene's encodes.
- Provenance, sources, and citations for data must be tracked more clearly. Was this data scraped or manually entered? What URLs can we find corroboration in? Did we pull this from StashDB in the first place? Was this pulled in as part of the starter pack? The current many2many relational tables (tags_images, performers_images, etc) need a new column for source of that data.
- Stash should access read-only stashdb out of the box, and users should be nudged to create an account to submit their edits. Editing your database should queue a change that, in a separate view, the user can either ignore or batch-submit to StashDB later.
- Scrapers should be one click to install. We already have them in a repo! Scraper list should be machine readable.
- StashDB should make regular database dumps available. This dataset represents the hard work of collectors, archivists, and enthusiasts. It should be maximally survivable and distributable. The gold standard for [data dumps is Wikipedia](https://dumps.wikimedia.org/backup-index.html), that makes all database tables public and regularly dumped to public fileservers. A monthly `pgdump` to a public directory will be enough, and can be an excellent starting point for a backup infrastructure. Will need to sanitize / exclude user table.

### ML Initiatives (3-5 year plan)
Not currently ready to leverage an ML pipeline yet, but with a little effort we can position ourselves to be ready to adopt it when it's time. The most important step is developing large collections of usefully annotated data. Two applications are immediately apparent:

- Associating a tag with a scene is near useless for this, but annotating a tag with a range of specific timestamps (like markers) can allow us to extract tagged frames for training. There is prior art in the form of [Miles Deep](https://github.com/ryanjay0/miles-deep), which tags sex positions in video. Annotated frames is how it was trained, and the method is generalizable to many (most?) types of tags. Recommendation: Markers should be allowed to overlap, should specify and end time, and should be uploaded to StashDB. We don't have to collect frame data just yet.
- Images and Scenes tagged with Performers can be scanned with face detection, have faces extracted, and have facial recognition models built on the results. Scenes with one performer don't require disambiguation, and can be used to build the first model. From then on, media with multiple performers can simply choose the closest match from existing performer facial data and add the match to the training set.

We would need to refine an architecture where we can (with user consent) collect training data from Stash clients derived from the actual source media, train the models, and then distribute the trained models back to Stash for privacy-preserving client-side inference.

**Comment by WithoutPants** [comment]
Until stashDB is more tightly integrated with the stashapp org overall (it's currently maintained by a single developer), I'm not keen on hard-coding the instance data into the source code. 

In the short-term, to encourage community involvement in stash-box instances - in this case specifically, stashDB - the setup wizard should include an option to setup a stash-box connection, with a link to a Github wiki page that provides links and potentially a read-only API key for endorsed instances.

**Comment by laurus-lx** [comment]
StashBox should be federated IMHO, with cross-referencing between instances. To assist with synchronizing instances and data sources - I'd propose setting up Adult Performers Authority - a sql or document-store db that will aggregate data across sources, and keep track of changes and reputation/data confidence that StashBox maintainers can selectively pull/push changes.
![image](https://user-images.githubusercontent.com/83609083/135908551-6f76b561-2244-416c-bf65-03edfc32c4e5.png)

**Comment by philpw99** [comment]
Just 2 cents from a newbie scraper like me:
It will be hard for everyone to agree with some information, like a performer's details, her ethnicity, and other fields... You will spend quite a lot of time collecting conflicting data, and people will keep arguing with it. It will comes to personal preference: some people will like the info from freeones.com, some people prefers Babepedia.com. Some will like the headshots from IAFD, and some will say the other website's headshot is way better...etc.
The porn database itself is highly personal already. It's not a good idea to set performers or scenes with the same data for everyone. I thought StashApp solved the problem very elegantly: you use your own scrapers. If you like Babepedia over Freeones, use the according scraper. Get the info you like, not the info that needs everyone's approval.
Therefore I think the key of StashApp at this stage, should be 3:
* Easy to install and run. For windows, it should come with a installer, and a proper GUI interface to run the server. I can think of an example: a program called "NZB Get" demonstrates this very nicely.
* Easy to use scrapers. Most people don't know that you need to copy those scrapers to the "scrapers" folder, in order to get the information from websites. The community scrapers should come with the installation, and they will be updated when the repo master get updated. Also it will be great if there is a simple way to enable/disable individual scrapers.
* Flexible when handling a scene file. Right now JW Player is the way to play, but there are so many other players that people use for their porn collection. If StashApp allows them to open the scene files with their own choice, like PotPlayer, VLC, or my favorite: DeoVR, the user experience will greatly improved. People will use Stash as the content manager, and still keep using their favorite players to view the files. I for one will be very happy.
If StashApp can accomplish these three, I am certain a lot more people will use it.

**Comment by Chodbin** [comment]
Another new user here, just over a week, adding my 2 cents

`As a new user I could be repeating things already mentioned or discussed elsewhere so I will not be offended if some or any of this is completely ignored`

## Growing Our Userbase
### Stash new user funnel: Finding and installing the Stash client
> Simplified installation instructions, de-emphasize or move to separate document developer info
> Graphics and shininess - Screenshots in the readme at the minimum

Base anything on the assumption that the new user is technologically illiterate and provide step by step guides

As @philpw99 said, a GUI interface would be nice, but in addition to web based rather than a replacement

### Stash new user funnel: First-use, and reduce time to usefulness
>We should attempt to open the default browser automatically after server start

Definitely, anything that reduces the number of actions (clicks, scrolls etc)  a user needs to make is an improvement, but could also be an option in configuration if people want to turn off for some reason

>Empty / 0 entry scene / image list should have stashes list and scan button
>Empty performers / studios / tags list should have option to auto-populate with starter from StashDB.

Before the user needs any of these, they need to have something in stash, is there a "best practice" for a new user?

`Should they scan all of there files? Selective scan some? generate phashes, sprites etc while scanning or after? What are phashes and sprites and what are they for? Should they create studios and performers first, or let stash create them when scraping?` 

Some of these are answered in help but maybe there should maybe be a welcome page that opens on first installation, that highlights the help information and has some kind of guide to getting started

**_Scrapers & Plugins_**
more from @philpw99 
> * Easy to use scrapers. Most people don't know that you need to copy those scrapers to the "scrapers" folder, in order to get the information from websites. The community scrapers should come with the installation, and they will be updated when the repo master get updated. Also it will be great if there is a simple way to enable/disable individual scrapers.

Ideally you should be able to install these directly from within stash, a section in top menu? (maybe also have a links to stashDB, Discord and Github in top menu?) For example, on the studio page, if there are scrapers available for studios that you don't have installed, there could be a link to install them

>StashDB should be pre-configured with a read-only user out-of-the-box. A user should be able to start up stash and start leveraging our data immediately, without joining our discord and messing with API keys. ("What's an API key? Do I have to pay for that? Plex just told me what my movies were automatically!")

100% agree

#### Starter Database: A symptom of missing features.
>StashDB should have a living set of "Starter" or "Common" data, perhaps implemented by a flag on each Performer, Tag, or Studio that denotes it as a good starter. (Criteria can be settled on, but anything that's extremely common or widely distributed is a good candidate) On first run, Stash should offer to automatically populate these directly from StashDB, and kick-off a scan and auto-tag job immediately after. This would get new users immediately started on a good foundation toward organization and a clear example of Stash's capabilities.

I see some issues with this, mainly that the user could end up with lots of unwanted/unneeded performers and tags, after all stash and the stashDB have no idea what might be in a users collection.
I think either stash should be able to detect from scanning what studios/performers/tags are missing and then offer to populate from that, or the user should be able to populate by importing from stashDB before scanning and matching.
For example, I have some Vixen network scenes, so i go to stashDB and import what i want/need, then i scan my Vixen Network folder, then i match them. Something like this maybe;

![StashDBStudioImport](https://user-images.githubusercontent.com/80072025/136685774-8b374560-d7cd-4da2-9232-e211345031fb.jpg)

There would probably have be some configuration options for what fields you want to import for the performers, As @philpw99 pointed out, peoples preferences vary. It's one of the best things about stashapp, the ability to customize and choose what you want to scrape and from where.

**_However_**

Whatever happens with using stashDB, it will need to be accurate and up to date, which imo, needs a lot more work. I'll go into that in more depth later. (never mind, I see a lot of it is covered in **Growing Our Dataset**)

### Make Stash a delight to use
>Stash has a few papercuts. After providing support to many, many users, I've distilled the most common sources of confusion:

For me it's only been a short while using stash, and i'm still in the process of building my collection rather than using it for viewing purposes but I've already noticed a lot of annoyances. (Don't get me wrong, I'm very impressed so far but think there is plenty of room for improvement)

# Ease of use
Whether on a mobile, tablet or desktop, fewer is better: The fewer actions needed to be taken (clicks, scrolls, pages visited, movement required [#1769 for example]) the better the experience and the quicker it is to achieve your aims.

For example, when viewing the studios, performers or scenes pages (I assume it's the same for all other sections) if i want to edit something, it requires 2 actions, first clicking on the item and then on the edit button. Having an edit option/button on each item would reduce this to 1 action. Even then, I need to perform 1 action to get back to the previews page to edit something else, ideally, clicking this edit option would open up a pop up like the help menu.

# Customization

Should be easy, manged from within stash and for the user to need no knowledge of CSS or coding. Maybe the ability to install custom themes from within stash for example? Personally I'd like to see the resizing bar that is available in the scenes/images sections also in the performers and studios sections.

There is a lot of wasted space when you select a studio (see image below), almost half of my screen, there must be a better way to do this...

![StudioSceneViewFullScreenHighlight](https://user-images.githubusercontent.com/80072025/136685787-1de4dec7-7ee7-401c-a215-3606036df4b5.jpg)

The search and filter options are powerful but not intuitive. When selecting a sort preference maybe stash could generate another navigation bar like the page number one;

- A-Z for relevant sorting like Name, Title
- Year and then month for Date, Birthdate, Created At
- Number of stars for rating etc

# Performer Aliases

I'd like to see better handling of performer aliases in stash (probably means better handling in stash DB as well) It's one of my major annoyances in scraping information and trying to organised my collection, and in fact it's the ability to edit this in stash that initially attracted me to it.

The information is out there on various sites, but would it would obviously be better if this was correct in stashDB. I've no idea about coding or how it is currently handled in stashDB but here is what i would like to see from the front end as a User

- I'd like to be able to choose what performer name/alias is used as the primary display name without this affecting any future scraping
- I'd like to see both the display name AND the alias when viewing information about a particular scene or studio (see image below for examples)
- This probably requires changes in the performer details page, either under Edit or another section specifically for Aliases (see 2nd image below for example)

![ScenePerformer3](https://user-images.githubusercontent.com/80072025/136685809-c032b59f-e6af-4dd3-a12c-1c41cc2b80fb.jpg)

![PerformerDetailsAliases](https://user-images.githubusercontent.com/80072025/136685818-1e244025-57c8-49fb-955d-87de799f3c70.jpg)

## Growing Our Dataset & Making Our Data Accessible and Useful

This is all great planning. 👍 It seems the long term plan is basically more integration (syncing?) between stash and stashDB? 

Short term I would like to see stashDB getting updated, I love the way the tagger works with the fingerprints and how easy it makes scraping info on scenes. I hate not being able to use it for scenes from the last 6 months or for studios not included. I have this frustration that my individual scraping and editing of scenes is duplicating the work everyone else is doing. I "think" work is going on backend? Difficult to find out exactly without searching through Discord or GitHub. For a new user who might only see stash and stashDB, no updates of the database for 6 months can look like it has been abandoned and could put new users off. There should probably be some kind of message about the status somewhere.

_On a side note, a link to stashapp on stashDB maybe?, just in case people come across that first_

I'd like to add a few suggestions/ideas;

We should try to avoid duplication of work if possible, by that I mean there is no point in several people editing the same studio for example, if there is another studio in desperate need of attention and no-one working on it. I think there should be some way to indicate that a studio, scene, performer, tags etc need attention/editing. Some Ideas;

### In stashDB

- An accuracy and/or completeness rating for studios, scenes, performers could indicate what needs most attention
   - This should probably be restricted to verifiable information rather than subjective like performer images or scene screenshots
- For users with editing privileges, some way of indicating what they are working on, a way to see what others are working on and a way to see what needs attention and is not being worked on.
- For users without editing privileges an option to mark items that require attention or to submit editing suggestions
   - For example, you are browsing a scene and see there is no performer attached, click a button to add a note/suggest an edit

### In stash
- An option to enable/disable syncing with stashDB to indicate what items in your collection (and in stashDB) need the most work (not everyone will want to contribute)
- The ability for Admins/Editors to push suggestions/requests for help on specific studios/scenes/movies/performers etc (option to disable)

### In Both
- Some way of adding old movies/scenes from defunct studios/websites
   - This should probably have some kind of notification/disclaimer that the content is purely user generated

## Other

- One thing I would like to see at some point in stash is the ability to compare my scenes, movies, performers or image sets with stashDB (or elsewhere) and then provide a list that are missing from my collection
- I haven't looked at movies yet but for scenes I'd like to be able to include both a poster and a cover, and then decide if i view the scenes by cover or poster
- Is Drag and Drop an option? For example drag and drop an image to edit, or drag a file on top of a studio to start scanning or scraping

Thanks for reading this far 👍 hopefully something of use among all that.

**Reply by Chodbin** [reply]
> For example, when viewing the studios, performers or scenes pages (I assume it's the same for all other sections) if i want to edit something, it requires 2 actions, first clicking on the item and then on the edit button. Having an edit option/button on each item would reduce this to 1 action. Even then, I need to perform 1 action to get back to the previews page to edit something else, ideally, clicking this edit option would open up a pop up like the help menu.

Reading this back, I realize I'm describing how Plex works.... 🤦

**Reply by kermieisinthehouse** [reply]
Good points and I'll address them individually in a longer post, but please immediately remove any media that isn't SFW, to comply with GitHub's ToS

**Reply by Chodbin** [reply]
ah ok, ill get on that right now sorry

**Reply by Chodbin** [reply]
> Good points and I'll address them individually in a longer post, but please immediately remove any media that isn't SFW, to comply with GitHub's ToS

Apologies, newbie mistake, ok now?
